mackerel-plugin-aws-kinesis-streams/lib/aws-kinesis-streams.go

```diff
@@ -0,0 +1,245 @@
+	mp "github.com/mackerelio/go-mackerel-plugin-helper"
+)
+
+const (
+	namespace          = "AWS/Kinesis"
+	metricsTypeAverage = "Average"
+	metricsTypeSum     = "Sum"
+	metricsTypeMaximum = "Maximum"
+	metricsTypeMinimum = "Minimum"
+)
+
+type metrics struct {
+	CloudWatchName string
+	MackerelName   string
+	Type           string
+}
+
+// KinesisStreamsPlugin mackerel plugin for aws kinesis
+type KinesisStreamsPlugin struct {
+	Name   string
+	Prefix string
+
+	AccessKeyID     string
+	SecretAccessKey string
+	Region          string
+	CloudWatch      *cloudwatch.CloudWatch
+}
+
+// MetricKeyPrefix interface for PluginWithPrefix
+func (p KinesisStreamsPlugin) MetricKeyPrefix() string {
+	if p.Prefix == "" {
+		p.Prefix = "kinesis-streams"
+	}
+	return p.Prefix
+}
+
+// prepare creates CloudWatch instance
+func (p *KinesisStreamsPlugin) prepare() error {
+	sess, err := session.NewSession()
+	if err != nil {
+		return err
+	}
+
+	config := aws.NewConfig()
+	if p.AccessKeyID != "" && p.SecretAccessKey != "" {
+		config = config.WithCredentials(credentials.NewStaticCredentials(p.AccessKeyID, p.SecretAccessKey, ""))
+	}
+	if p.Region != "" {
+		config = config.WithRegion(p.Region)
+	}
+
+	p.CloudWatch = cloudwatch.New(sess, config)
+
+	return nil
+}
+
+// getLastPoint fetches a CloudWatch metric and parse
+func (p KinesisStreamsPlugin) getLastPoint(metric metrics) (float64, error) {
+	now := time.Now()
+
+	dimensions := []*cloudwatch.Dimension{
+		{
+			Name:  aws.String("StreamName"),
+			Value: aws.String(p.Name),
+		},
+	}
+
+	response, err := p.CloudWatch.GetMetricStatistics(&cloudwatch.GetMetricStatisticsInput{
+		Dimensions: dimensions,
+		StartTime:  aws.Time(now.Add(time.Duration(180) * time.Second * -1)), // 3 min
+		EndTime:    aws.Time(now),
+		MetricName: aws.String(metric.CloudWatchName),
+		Period:     aws.Int64(60),
+		Statistics: []*string{aws.String(metric.Type)},
+		Namespace:  aws.String(namespace),
+	})
+	if err != nil {
+		return 0, err
+	}
+
+	datapoints := response.Datapoints
+	if len(datapoints) == 0 {
+		return 0, errors.New("fetched no datapoints")
+	}
+
+	latest := new(time.Time)
+	var latestVal float64
+	for _, dp := range datapoints {
+		if dp.Timestamp.Before(*latest) {
+			continue
+		}
+
+		latest = dp.Timestamp
+		switch metric.Type {
+		case metricsTypeAverage:
+			latestVal = *dp.Average
+		case metricsTypeSum:
+			latestVal = *dp.Sum
+		case metricsTypeMaximum:
+			latestVal = *dp.Maximum
+		case metricsTypeMinimum:
+			latestVal = *dp.Minimum
+		}
+	}
+
+	return latestVal, nil
+}
+
+// FetchMetrics fetch the metrics
+func (p KinesisStreamsPlugin) FetchMetrics() (map[string]interface{}, error) {
+	stat := make(map[string]interface{})
+
+	for _, met := range [...]metrics{
+		{CloudWatchName: "GetRecords.Bytes", MackerelName: "GetRecordsBytes", Type: metricsTypeAverage},
+		{CloudWatchName: "GetRecords.IteratorAgeMilliseconds", MackerelName: "GetRecordsDelayMaxMilliseconds", Type: metricsTypeMaximum},
```

Comment:
`GetRecords.IteratorAgeMilliseconds` fetches max and min values, but others fetch only average value. It would be better to have some comments why you chose max and min values for this metric.